File: test_bare_bones.py
#!/usr/bin/env python3
"""
Bare-bones Drip Python SDK test.
Follows README docs exactly. No mocks, no frameworks, just real API calls.

Usage:
    python test_bare_bones.py
"""
import os
import sys
import time

# Step 1: Test .env loading (README says: "auto-loads .env files when python-dotenv is installed")
# We do NOT manually call load_dotenv() — the SDK should handle it.

from drip import Drip, DripError, DripAPIError

PASS = 0
FAIL = 0

def check(name, fn):
    global PASS, FAIL
    try:
        result = fn()
        print(f"  PASS  {name}")
        PASS += 1
        return result
    except Exception as e:
        print(f"  FAIL  {name}: {e}")
        FAIL += 1
        return None


print("=" * 60)
print("Drip Python SDK — Bare Bones Test")
print("=" * 60)

# --- SDK INIT ---
# README: "Auto-reads DRIP_API_KEY from environment"
# README: "Or pass config explicitly"
print("\n[1] SDK Initialization")

client = None
def init_from_env():
    global client
    # README says this should work if DRIP_API_KEY is in env or .env
    client = Drip()
    assert client is not None
    return client

check("Drip() reads DRIP_API_KEY from env/.env", init_from_env)

if client is None:
    print("\nFATAL: SDK could not initialize. Is DRIP_API_KEY set in .env or environment?")
    sys.exit(1)

print(f"  INFO  key_type = {client._key_type}")
print(f"  INFO  base_url = {client._base_url}")

# --- PING ---
# README: "drip.ping() — Verify API connection"
print("\n[2] Ping")
check("ping()", lambda: client.ping())

# --- CREATE CUSTOMER ---
# README: "customer = drip.create_customer(external_customer_id='user_123')"
print("\n[3] Create Customer")
ext_id = f"barebone-test-{int(time.time())}"
customer = check("create_customer(external_customer_id=...)",
    lambda: client.create_customer(external_customer_id=ext_id))

if customer is None:
    print("\nFATAL: Could not create customer. Stopping.")
    sys.exit(1)

print(f"  INFO  customer.id = {customer.id}")

# --- GET CUSTOMER ---
# README: "get_customer(customer_id)"
print("\n[4] Get Customer")
fetched = check("get_customer(customer_id)",
    lambda: client.get_customer(customer.id))

# --- LIST CUSTOMERS ---
# README: "list_customers(options)"
print("\n[5] List Customers")
listing = check("list_customers()", lambda: client.list_customers())
if listing:
    print(f"  INFO  count = {listing.count}")

# --- TRACK USAGE ---
# README: "drip.track_usage(customer_id=customer.id, meter='api_calls', quantity=1)"
print("\n[6] Track Usage")
usage_result = check("track_usage(customer_id, meter, quantity)",
    lambda: client.track_usage(
        customer_id=customer.id,
        meter="api_calls",
        quantity=1
    ))

# --- TRACK USAGE WITH METADATA ---
# README: "drip.track_usage(..., metadata={'model': 'gpt-4o-mini'})"
print("\n[7] Track Usage with Metadata")
check("track_usage(..., metadata={...})",
    lambda: client.track_usage(
        customer_id=customer.id,
        meter="llm_tokens",
        quantity=842,
        metadata={"model": "gpt-4o-mini"}
    ))

# --- CHARGE ---
# README: "drip.charge(customer_id=customer.id, meter='api_calls', quantity=1)"
print("\n[8] Charge")
charge_result = check("charge(customer_id, meter, quantity)",
    lambda: client.charge(
        customer_id=customer.id,
        meter="api_calls",
        quantity=1
    ))
if charge_result:
    print(f"  INFO  charge = {charge_result}")

# --- CHARGE ASYNC ---
# README: "charge_async(customer_id, meter, quantity) — returns immediately"
print("\n[9] Charge Async")
async_result = check("charge_async(customer_id, meter, quantity)",
    lambda: client.charge_async(
        customer_id=customer.id,
        meter="api_calls",
        quantity=1
    ))
if async_result:
    print(f"  INFO  async_result = {async_result}")

# --- LIST CHARGES ---
# README: "list_charges(options)"
print("\n[10] List Charges")
charges = check("list_charges()", lambda: client.list_charges())
if charges:
    print(f"  INFO  count = {charges.count}")

# --- GET BALANCE ---
# README: "get_balance(customer_id)"
print("\n[11] Get Balance")
balance = check("get_balance(customer_id)",
    lambda: client.get_balance(customer.id))
if balance:
    print(f"  INFO  balance = {balance}")

# --- RECORD RUN ---
# README: "drip.record_run(customer_id=..., workflow='research-agent', events=[...], status='COMPLETED')"
print("\n[12] Record Run")
run_result = check("record_run(customer_id, workflow, events, status)",
    lambda: client.record_run(
        customer_id=customer.id,
        workflow="research-agent",
        events=[
            {"event_type": "llm.call", "quantity": 1700, "units": "tokens"},
            {"event_type": "tool.call", "quantity": 1},
        ],
        status="COMPLETED"
    ))
if run_result:
    print(f"  INFO  run = {run_result}")

# --- RUN CONTEXT MANAGER ---
# README: "with drip.run('research-agent', customer_id=customer.id) as run:"
print("\n[13] Run Context Manager")
def test_run_ctx():
    with client.run("ctx-test-agent", customer_id=customer.id) as run:
        run.event("llm.call", quantity=500, units="tokens")
        run.event("tool.call", quantity=1)
    return True
check("run() context manager", test_run_ctx)

# --- LIST EVENTS ---
# README: "list_events(options)"
print("\n[14] List Events")
events = check("list_events()", lambda: client.list_events())
if events:
    print(f"  INFO  events count = {len(events.data) if hasattr(events, 'data') else events}")

# --- ERROR HANDLING ---
# README: "except DripAPIError as e: print(f'API error {e.status_code}: {e.message}')"
print("\n[15] Error Handling")
def test_error():
    try:
        client.get_customer("nonexistent-id-12345")
        return False  # Should have thrown
    except DripAPIError as e:
        print(f"  INFO  Caught DripAPIError: status={e.status_code} message={e.message}")
        return True
    except DripError as e:
        print(f"  INFO  Caught DripError: {e}")
        return True
check("DripAPIError on 404", test_error)

# --- SINGLETON ---
# README: "from drip import drip" then "drip.ping()"
print("\n[16] Singleton")
def test_singleton():
    from drip import drip as drip_singleton
    drip_singleton.ping()
    return True
check("drip singleton auto-init", test_singleton)

# --- SUMMARY ---
print("\n" + "=" * 60)
print(f"RESULTS: {PASS} passed, {FAIL} failed out of {PASS + FAIL}")
print("=" * 60)

if FAIL > 0:
    sys.exit(1)

File: test_bare_bones.mjs
/**
 * Bare-bones Drip Node.js SDK test.
 * Follows README docs exactly. No mocks, no frameworks, just real API calls.
 *
 * Usage:
 *     node test_bare_bones.mjs
 */

// NOTE: SDK README says "auto-loads .env when dotenv is installed" but this
// doesn't work in practice because tsup tree-shakes the require('dotenv') call.
// You MUST load dotenv yourself. This is a documentation bug.
import 'dotenv/config';

import { Drip, DripError } from '@drip-sdk/node';

let PASS = 0;
let FAIL = 0;

async function check(name, fn) {
  try {
    const result = await fn();
    console.log(`  PASS  ${name}`);
    PASS++;
    return result;
  } catch (e) {
    console.log(`  FAIL  ${name}: ${e.message || e}`);
    FAIL++;
    return null;
  }
}

console.log('='.repeat(60));
console.log('Drip Node.js SDK — Bare Bones Test');
console.log('='.repeat(60));

// --- SDK INIT ---
// README: "Auto-reads DRIP_API_KEY from environment"
console.log('\n[1] SDK Initialization');

let client;
try {
  // README says: const client = new Drip();
  client = new Drip();
  console.log(`  PASS  new Drip() reads DRIP_API_KEY from env/.env`);
  PASS++;
} catch (e) {
  console.log(`  FAIL  new Drip(): ${e.message}`);
  FAIL++;
  console.log('\nFATAL: SDK could not initialize. Is DRIP_API_KEY set in .env or environment?');
  process.exit(1);
}

console.log(`  INFO  keyType = ${client.keyType}`);
console.log(`  INFO  baseUrl = ${client.baseUrl}`);

// --- PING ---
// README: "await drip.ping()"
console.log('\n[2] Ping');
await check('ping()', () => client.ping());

// --- CREATE CUSTOMER ---
// README: "const customer = await drip.createCustomer({ externalCustomerId: 'user_123' })"
console.log('\n[3] Create Customer');
const extId = `barebone-test-${Date.now()}`;
const customer = await check('createCustomer({ externalCustomerId })',
  () => client.createCustomer({ externalCustomerId: extId }));

if (!customer) {
  console.log('\nFATAL: Could not create customer. Stopping.');
  process.exit(1);
}
console.log(`  INFO  customer.id = ${customer.id}`);

// --- GET CUSTOMER ---
// README: "getCustomer(customerId)"
console.log('\n[4] Get Customer');
const fetched = await check('getCustomer(customerId)',
  () => client.getCustomer(customer.id));

// --- LIST CUSTOMERS ---
// README: "listCustomers(options)"
console.log('\n[5] List Customers');
const listing = await check('listCustomers()', () => client.listCustomers());
if (listing) console.log(`  INFO  count = ${listing.count}`);

// --- TRACK USAGE ---
// README: "await drip.trackUsage({ customerId: customer.id, meter: 'api_calls', quantity: 1 })"
console.log('\n[6] Track Usage');
const usageResult = await check('trackUsage({ customerId, meter, quantity })',
  () => client.trackUsage({
    customerId: customer.id,
    meter: 'api_calls',
    quantity: 1,
  }));

// --- TRACK USAGE WITH METADATA ---
// README: "await drip.trackUsage({ ..., metadata: { model: 'gpt-4o-mini' } })"
console.log('\n[7] Track Usage with Metadata');
await check('trackUsage({ ..., metadata })',
  () => client.trackUsage({
    customerId: customer.id,
    meter: 'llm_tokens',
    quantity: 842,
    metadata: { model: 'gpt-4o-mini' },
  }));

// --- CHARGE ---
// README: "await drip.charge({ customerId, meter, quantity })"
console.log('\n[8] Charge');
const chargeResult = await check('charge({ customerId, meter, quantity })',
  () => client.charge({
    customerId: customer.id,
    meter: 'api_calls',
    quantity: 1,
  }));
if (chargeResult) console.log(`  INFO  charge = ${JSON.stringify(chargeResult).slice(0, 100)}`);

// --- CHARGE ASYNC ---
// README: "chargeAsync(params) — returns 202 immediately"
console.log('\n[9] Charge Async');
const asyncResult = await check('chargeAsync({ customerId, meter, quantity })',
  () => client.chargeAsync({
    customerId: customer.id,
    meter: 'api_calls',
    quantity: 1,
  }));
if (asyncResult) console.log(`  INFO  asyncResult = ${JSON.stringify(asyncResult).slice(0, 100)}`);

// --- LIST CHARGES ---
// README: "listCharges(options?)"
console.log('\n[10] List Charges');
const charges = await check('listCharges()', () => client.listCharges());
if (charges) console.log(`  INFO  count = ${charges.count}`);

// --- GET BALANCE ---
// README: "getBalance(customerId)"
console.log('\n[11] Get Balance');
const balance = await check('getBalance(customerId)',
  () => client.getBalance(customer.id));
if (balance) console.log(`  INFO  balance = ${JSON.stringify(balance)}`);

// --- RECORD RUN ---
// README: "await drip.recordRun({ customerId, workflow, events, status })"
console.log('\n[12] Record Run');
const runResult = await check('recordRun({ customerId, workflow, events, status })',
  () => client.recordRun({
    customerId: customer.id,
    workflow: 'research-agent',
    events: [
      { eventType: 'llm.call', quantity: 1700, units: 'tokens' },
      { eventType: 'tool.call', quantity: 1 },
    ],
    status: 'COMPLETED',
  }));
if (runResult) console.log(`  INFO  run = ${JSON.stringify(runResult).slice(0, 100)}`);

// --- LIST EVENTS ---
// README: "listEvents(options?)"
console.log('\n[13] List Events');
const events = await check('listEvents()', () => client.listEvents());
if (events) console.log(`  INFO  events count = ${events.data?.length ?? 'unknown'}`);

// --- ERROR HANDLING ---
// README: "if (error instanceof DripError) { console.error(...) }"
console.log('\n[14] Error Handling');
await check('DripError on 404', async () => {
  try {
    await client.getCustomer('nonexistent-id-12345');
    throw new Error('Should have thrown');
  } catch (e) {
    if (e instanceof DripError) {
      console.log(`  INFO  Caught DripError: code=${e.code} message=${e.message}`);
      return true;
    }
    // Check if it's a generic Error from the SDK
    if (e.message && e.message.includes('404')) {
      console.log(`  INFO  Caught Error with 404: ${e.message}`);
      return true;
    }
    throw e;
  }
});

// --- SINGLETON ---
// README: "import { drip } from '@drip-sdk/node'"
console.log('\n[15] Singleton');
await check('drip singleton auto-init', async () => {
  const { drip } = await import('@drip-sdk/node');
  await drip.ping();
  return true;
});

// --- SUMMARY ---
console.log('\n' + '='.repeat(60));
console.log(`RESULTS: ${PASS} passed, ${FAIL} failed out of ${PASS + FAIL}`);
console.log('='.repeat(60));

if (FAIL > 0) process.exit(1);
